extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLSubscriptionHandler.java

```diff
@@ -0,0 +1,180 @@
+        this.config = config;
+        this.executionErrorsService = new ExecutionErrorsService(config);
+    }
+
+    @Override
+    protected void doHandle(final RoutingContext ctx) {
+        if (ctx.request().headers().contains(HttpHeaders.UPGRADE, HttpHeaders.WEBSOCKET, true) && !ctx.request().isEnded()) {
+            ctx.request().toWebSocket(new SmallRyeWebSocketHandler());
+        } else {
+            ctx.next();
+        }
+    }
+
+    private class SmallRyeWebSocketHandler implements Handler<AsyncResult<ServerWebSocket>> {
+
+        @Override
+        public void handle(AsyncResult<ServerWebSocket> event) {
+            if (event.succeeded()) {
+                ServerWebSocket serverWebSocket = event.result();
+                serverWebSocket.closeHandler(new CloseHandler());
+                serverWebSocket.endHandler(new EndHandler());
+                serverWebSocket.exceptionHandler(new ExceptionHandler());
+                serverWebSocket.textMessageHandler(new TextMessageHandler(serverWebSocket));
+            }
+        }
+    }
+
+    private class CloseHandler implements Handler<Void> {
+        @Override
+        public void handle(Void e) {
+            unsubscribe();
+        }
+    }
+
+    private class EndHandler implements Handler<Void> {
+        @Override
+        public void handle(Void e) {
+            unsubscribe();
+        }
+    }
+
+    private class ExceptionHandler implements Handler<Throwable> {
+        @Override
+        public void handle(Throwable e) {
+            log.error(e.getMessage());
+            unsubscribe();
+        }
+    }
+
+    public void unsubscribe() {
+        Subscriptions.cancel(subscriptionRef);
+        subscriptionRef.set(null);
+    }
+
+    private class TextMessageHandler implements Handler<String> {
+        private final SmallRyeGraphQLSubscriptionSubscriber smallRyeGraphQLSubscriptionSubscriber;
+
+        TextMessageHandler(final ServerWebSocket serverWebSocket) {
+            this.smallRyeGraphQLSubscriptionSubscriber = new SmallRyeGraphQLSubscriptionSubscriber(serverWebSocket);
+        }
+
+        @Override
+        public void handle(String message) {
+            JsonObject jsonInput = inputToJsonObject(message);
+            ExecutionResponse executionResponse = getExecutionService()
+                    .execute(jsonInput);
+
+            ExecutionResult executionResult = executionResponse.getExecutionResult();
+            if (executionResult != null) {
+                // If there is error on the query, we can not start a subscription
+                if (executionResult.getErrors() != null && !executionResult.getErrors().isEmpty()) {
+                    smallRyeGraphQLSubscriptionSubscriber.onNext(executionResult);
+                    smallRyeGraphQLSubscriptionSubscriber.closeWebSocket();
+                } else {
+                    Publisher<ExecutionResult> stream = executionResponse.getExecutionResult()
+                            .getData();
```

Comment:
Why isn't this method returning a `Multi` directly?

Reply:
This comes from graphql-java (see https://www.graphql-java.com/documentation/v16/subscriptions/) that support Publisher. This is the 'final' result from the query, that included multiple dataloaders (including the root one of Multi, but also fields and possibly Source).  At the moment I am making that a Multi, so that I can add the onFailure, but I am busy moving that to SmallRye.